File: ts/src/cache/lru.ts
import { CacheElement } from "./type/cache.type";

export class LRUCache<K, T> {
  private dataArray: Array<CacheElement<K, T>> = [];
  private maxSize: number;
  private eqFn: (k1: K, k2: K) => boolean;

  constructor(
    size: number,
    eqFn: (k1: K, k2: K) => boolean = (k1, k2) => k1 === k2,
  ) {
    this.maxSize = size;
    this.eqFn = eqFn;
  }

  public set(key: K, value: T) {
    const index = this.dataArray.findIndex(p => this.eqFn(p.key, key));
    const data =
      index > -1 ? this.dataArray.splice(index, 1)[0] : { key, value };
    if (this.dataArray.length === this.maxSize - 1) {
      this.dataArray.shift();
    }
    this.dataArray.push(data);
  }

  public get(key: K) {
    const index = this.dataArray.findIndex(p => this.eqFn(p.key, key));
    if (index > -1) {
      const pair = this.dataArray.splice(index, 1)[0];
      this.dataArray.push(pair);
      return pair.value;
    }
    return undefined;
  }
}

File: python/src/cache/lru.py
from typing import TypeVar, Callable, List

K = TypeVar('K')
T = TypeVar('T')


class CacheElement:
    def __init__(self, key: K, value: T):
        self.key = key
        self.value = value


class LRUCache:
    def __init__(self, size: int, eq_fn: Callable[[K, K], bool] = lambda k1, k2: k1 == k2):
        self.data_array: List[CacheElement] = []
        self.max_size = size
        self.eq_fn = eq_fn

    def set(self, key: K, value: T) -> None:
        index = next((i for i, p in enumerate(self.data_array)
                     if self.eq_fn(p.key, key)), -1)
        data = CacheElement(key, value)
        if index > -1:
            self.data_array.pop(index)
        elif len(self.data_array) == self.max_size - 1:
            self.data_array.pop(0)
        self.data_array.append(data)

    def get(self, key: K) -> T:
        index = next((i for i, p in enumerate(self.data_array)
                     if self.eq_fn(p.key, key)), -1)
        if index > -1:
            pair = self.data_array.pop(index)
            self.data_array.append(pair)
            return pair.value
        return None
